Kafka Connect Mongo Depositor

Michael7371 opened the pull request:
This PR includes the following:

- Kafka Connect service with an initial configuration for MongoDB. It can also connect to S3, GCS, etc ([Brief Description of Kafka Connect](https://www.confluent.io/blog/kafka-connect-tutorial/))
- Example config file includes single message transforms to add a MongoDB recognized datetime field at the root to use for TTL indexes or other queries. Also adds support for converting ID fields to be used as the document ID when it is stored in Mongo.

Michael7371 committed:
initial commit with mongo connector setup

Michael7371 committed:
updates to readme

Michael7371 committed:
updates to readme and made the sample_connect_start.sh contain ODE sinks.

Michael7371 committed:
addressing pr comments, adding documentation.

dmccoystephenson committed:
Added image documentation to `docs/dockerhub.md`

mwodahl committed:
Update Dockerfile, pom.xml

mwodahl committed:
Add log4j.properties

Michael7371 committed:
updates from testing

Michael7371 committed:
parameterized mongodb name

payneBrandon committed:
Merge pull request #9 from CDOT-CV/mongo-depositor

MongoDB Connector

payneBrandon committed:
Merge pull request #12 from CDOT-CV/Fix/log4j-initialize

Fix log4j initialization issue

payneBrandon committed:
Merge pull request #11 from CDOT-CV/Update/java-version

Update Java Version

mwodahl committed:
Update AwsDepositor.java to remove command-line variables

mwodahl committed:
Update run.sh

mwodahl committed:
Dockerfile + docker-compose file updates

dmccoystephenson committed:
Modified 'purpose' section in `dockerhub.md`

payneBrandon committed:
Merge pull request #10 from CDOT-CV/docs/dockerhub-description

DockerHub Image Documentation

dan-du-car commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM maven:3.5.4-jdk-8-alpine as builder
+FROM maven:3.8-eclipse-temurin-21-alpine as builder
 MAINTAINER 583114@bah.com
```
The maintainer instruction is deprecated, please remove it or replace it with LABEL. details:https://docs.docker.com/engine/reference/builder/#maintainer-deprecated ?

dan-du-car commented on .gitignore:
```diff
@@ -1,12 +1,15 @@
+/*.log
+
+.env
+connect_start.sh
```
Do not see this connect_start.sh.

Michael7371 replied:
I am planning on re-working this, it will be updated after I can get the changes merged into the CDOT fork.

dan-du-car commented on src/main/resources/log4j.properties:
```diff
@@ -0,0 +1,10 @@
+# Set the log level for AwsDepositor
+log4j.logger.us.dot.its.jpo.ode.aws.depositor.AwsDepositor=WARN
+log4j.additivity.us.dot.its.jpo.ode.aws.depositor.AwsDepositor=false
+log4j.appender.awsDepositor.threshold=DEBUG
```
What is the difference between the log4j WARN logger and DEBUG appender?

dan-du-car commented on mongo-connector/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+# Note that this image requires a version of Mongodb of 3.6 or later
+FROM confluentinc/cp-kafka-connect:6.1.9
+
+COPY connect_start.sh /scripts/connect_start.sh
```
Why do we choose the name sample_connect_start.sh rather than connect_start.sh? Also if sample_connect_start.sh is the preferred name outside container. Maybe update the copy command: 
```
COPY sample_connect_start.sh /scripts/connect_start.sh
```

Michael7371 replied:
The original idea is that the sample_connect_start.sh would be a starting point to create a connect_start.sh. This seems slightly confusing after further thought, and instead I am going to push some changes that allow for attaching different connect_start.sh files via environmental variables rather than embedding it in the image.

dmccoystephenson committed:
Merge branch 'usdot-jpo-ode:develop' into dev

dmccoystephenson committed:
Added release notes for v1.4.0

dan-du-car commented on docker-compose-mongo.yml:
```diff
@@ -0,0 +1,88 @@
+    environment:
+      MONGO_URI: mongodb://${MONGO_IP}:${MONGO_PORT}/?replicaSet=rs0
+      MONGO_DB_NAME: ODE
+      CONNECT_BOOTSTRAP_SERVERS: ${DOCKER_HOST_IP}:9092
```
Is this MONGO_IP always the same as the DOCKER_HOST_IP? Maybe use one variable if yes. I got this issue testing the connector in local:
![image](https://github.com/usdot-jpo-ode/jpo-s3-deposit/assets/62157949/b366ca0b-9e31-416a-a565-8cd944f0cd0e)

![image](https://github.com/usdot-jpo-ode/jpo-s3-deposit/assets/62157949/79f2ce7b-17db-4857-b829-64a5d944b64a)

It cannot connect to Kafka message broker.

Michael7371 replied:
In some cases, we do want to be able to vary the mongo_ip for more production deployments where a true replica-set mongodB deployment is desired.
I was able to reproduce the error you are seeing when I attempted to create connectors that didn't have a kafka topic already made for them. Can you try updating the KAFKA_CREATE_TOPICS variable (line 16 in docker-compose.yml) to the following:

KAFKA_CREATE_TOPICS: "topic.OdeSpatJson:1:1,topic.OdeBsmJson:1:1,topic.OdeTimJson:1:1,topic.OdePsmJson:1:1,topic.OdeRawEncodedPsmJson,topic.OdeRawEncodedBSMJson:1:1,topic.OdeRawEncodedSPATJson:1:1,topic.OdeRawEncodedTIMJson:1:1,topic.OdeRawEncodedMAPJson:1:1,topic.OdeMapJson:1:1,topic.OdeRawEncodedPSMJson:1:1,topic.OdePsmJson:1:1"

Michael7371 replied:
I have added this fix to a current PR in the CDOT fork.

dan-du-car replied:
Updated the create topics list. Still in the loop waiting for 200. Also tried curl command in host machine, I got this result:
```
dev@dev:~$ curl http://connect:8083/connectors
curl: (6) Could not resolve host: connect
```
This is the env settings:
```
DOCKER_HOST_IP=localhost
# MONGODB Variables
MONGO_IP=${DOCKER_HOST_IP}
MONGO_PORT=27017

```

Michael7371 replied:
Oh I see, localhost will not work. This will cause the kafka connect image to try to look at its own 9092 port for the kafka broker instead of through your host's network. Try runing 'ifconfig' in a wsl terminal and use the docker0 network or eth0 network IP.

dan-du-car replied:
I tried the docker0 IP, and it connected to Kafka message broker. However, the PUT request from the .sh script return 400 response multiple times within duration of 2 mins after containers are launched. 
![image](https://github.com/usdot-jpo-ode/jpo-s3-deposit/assets/62157949/fc0b6b08-1540-44c8-921c-05386ddcde33)

Michael7371 replied:
Interesting, can you share your current .env file and run the following commands for me in a WSL terminal and let me know the results:
sudo apt-get update
sudo apt-get install kafkacat
kafkacat -b localhost:9092 -G test_consumer_group topic.OdeSpatJson -f '\nTopic: %t\nKey: %k\nTimestamp: %T\nValue: %s\n' -o beginning

install mongosh in wsl/windows: https://www.mongodb.com/docs/mongodb-shell/install/
mongosh "mongodb://localhost:27017/?DirectConnection=true"

Also, the changes mentioned previously have now been merged into this PR so if you pull, your .env must be updated with the sample.env changes.

dan-du-car replied:
I am using ubuntu 22.04 Jammy distribution not windows.  Installed KafkaCat.
Pulled the latest dev branch and ran the command to build images and launch containers:  
` docker-compose -f docker-compose-mongo.yml up`
It seems up, then bring it down and up again. I got below issues:
![image](https://github.com/usdot-jpo-ode/jpo-s3-deposit/assets/62157949/9124d08c-6777-4df1-8304-bc7c00f0f3a5)

Michael7371 replied:
Okay, Ubuntu should work as well. The Mongodb password is required to set in your .env file. The docker0 network looks like the correct network to use.

Michael7371 replied:
Also now that the kafka connect image has changed please run the following command to bring the containers down and back up again:

docker-compose -f docker-compose-mongo.yml down -v
docker-compose -f docker-compose-mongo.yml up --build --force-recreate --no-deps -d

dan-du-car replied:
![image](https://github.com/usdot-jpo-ode/jpo-s3-deposit/assets/62157949/855a21c6-f076-404a-b94f-571f2532f990)

Michael7371 replied:
@dan-du-car We have just pushed some updates that should fix that exit code and other related issues.

dan-du-car replied:
Look good. Thank you!

Michael7371 committed:
Update Dockerfile and environment variables
- reworked the kafka connect image to be attached as a volume
- added initialization scripts to the mongo container to auto create TTL indexes
- added more mongodb variables to allow for more customization.

Michael7371 committed:
Update docker-compose-mongo.yml and mongo-connector scripts to allow for authentication

Michael7371 committed:
Fix typo in README.md

dmccoystephenson committed:
Changed version to 1.4.0-SNAPSHOT

drewjj committed:
Merge pull request #14 from CDOT-CV/docs/release-notes-for-1.4.0

Release Notes for v1.4.0

mwodahl committed:
Update to use system.getenv

dmccoystephenson committed:
Merge pull request #16 from CDOT-CV/version/change-version-to-1.4.0-SNAPSHOT

Changed version to 1.4.0-SNAPSHOT

mwodahl committed:
Remove dockerfile env var

payneBrandon committed:
Merge pull request #15 from CDOT-CV/mongo-depositor

Update Dockerfile and environment variables for Kafka Connect

mwodahl committed:
Merge branch 'dev' into Update/env_refactor

Michael7371 committed:
fixing script file permissions and kafka topic creation issues.

Michael7371 committed:
Update README.md with instructions for connecting to MongoDB and debugging Kafka connect image crash

dmccoystephenson committed:
Merge pull request #17 from CDOT-CV/mongo-script-permission-fix

Mongo Connector Fixes

mwodahl committed:
env variable update

payneBrandon committed:
Merge pull request #13 from CDOT-CV/Update/env_refactor

Remove command line arguments from S3-Depositor
